src/StringCodeUtilities.hpp

```diff
@@ -0,0 +1,100 @@
+NOTE 2: The idCode identifiers should uniquely identify a certain type among all derived types of StringCodeComponent.
+NOTE 3: However you may have multiple codes of the same type / identifier in a list (so use the index argument to distinguish).
+NOTE 4: The parameter identifiers of a class and its' derived types must be unique (just as the actual parameter names in code), so a params code list will always have unique element identifiers.
+NOTE 5: You do not have to leave out empty brackets ( ) or { } like in the examples above. Passing empty brackets is completely legal (just as passing empty codes, which then always yields empty function results).
+NOTE 6: Unfortunately parameters of string type must not contain any spaces, commas, or brackets of type () or {}.
+*/
+
+// --- Readers
+
+string readIdCode(const string &fullCode); // read idCode string from fullCode or treeCode
+string readParams(const string &fullCode); // read params string from fullCode or treeCode
+string readParamValue(const string &params, const string &paramIdCode); // return the value string of certain paramId
+string readMemberTreeCode(const string &treeCode); // return a list string composed of the treeCodes of all members in treeCode
+string readTreeCode(const string &memberTreeCode, const int &index, const string &memberIdCode = ""); // return the treeCode of the '(index-1)'th member (first member per default) from memberTreeCode (if passed, only those with matching memberIdCode identifier)
+
+// --- Drop
+
+string dropParams(const string &code); // returns the a copy of the same code with all params dropped (i.e. only idCodes)
+string dropMembers(const string &code, const int &drop_lvl = 1); // returns the a copy of the same code with all members after tree level lvl dropped (i.e. lvl==1 -> no members at all)
+
+// --- Counters
+
+int countNParams(const string &params); // count number of params in params string
+int countTreeNParams(const string &treeCode); // count total number of params in treeCode string
+int countNMembers(const string &memberTreeCode, const bool &direct_only = true); // count number of direct (or total if direct_only == false) members in memberTreeCode string
+
+// --- Composers
+
+string composeCodes(const string &code1, const string &code2); // compose a string of the two codes separated by spaces and comma
+string composeCodeList(const vector<string> &codes); // compose a string of vector elements separated by spaces and comma
+string composeFullCode(const string &idCode, const string &params); // compose fullCode string from idCode and params
+string composeTreeCode(const string &fullcode, const string &memberTreeCode); // compose treeCode string from fullCode and memberTreeCode
+
+
+// --- Templates
+
+// for applying parameter value string to actual parameter, from param value string
+template <typename T>
```

Comment:
I have been told that template slow down significantly the execution.
Now, I don't think that these functions are called very often when doing actual simulation so it probably doesn't matter. But I am not sure yet, so I post this comment.
Just confirm that it is not relevant

Reply:
I have been told kind of the opposite :-P
It will mainly slow down compilation of the code, because the compiler will generate different versions of this function for all the datatypes which get used in place of typename T. But then it is quite the same as writing down the functions yourself for each datatype. BTW, basically the whole C++ standard library works on basis of templates (STL).